TLS routing ping for database protocols

gabrielcorado opened the pull request:
Implements [RFD 81](https://github.com/gravitational/teleport/blob/master/rfd/0081-tls-ping.md).

gabrielcorado committed:
feat: ping connection for database protocols

greedy52 commented on lib/srv/alpnproxy/proxy.go:
```diff
@@ -379,14 +394,54 @@ func (p *Proxy) handleConn(ctx context.Context, clientConn net.Conn) error {
 
+	var handlerConn net.Conn = tlsConn
+	// Check if Multiplex is supported/required by the client.
+	if len(hello.SupportedProtos) > 0 && common.IsPingProtocol(common.Protocol(hello.SupportedProtos[0])) {
```
`getHandleDescBasedOnALPNVal` goes through the whole list so `hello.SupportedProtos[0]` here may not match the selected one

gabrielcorado replied:
I've updated this to use the negotiated protocol instead of the protocol list the client provided. Using it guarantees that both sides of the connection are using ping connections.

greedy52 reviewed (commented):
I really like the approach. It's a lot lighter than HTTP2 frame

gabrielcorado commented on lib/srv/alpnproxy/common/protocols.go:
```diff
@@ -129,9 +138,56 @@ func ToALPNProtocol(dbProtocol string) (Protocol, error) {
+
+// ProtocolsWithPingSupport is the list of protocols that Ping connection is
+// supported.
+var ProtocolsWithPingSupport = []Protocol{
+	ProtocolPostgres,
+	ProtocolMySQL,
+	ProtocolMongoDB,
+	ProtocolRedisDB,
+	ProtocolSQLServer,
+	ProtocolSnowflake,
+}
```
I kept it in a separate definition since I wasn't sure if we wanted to add other protocols (for example, application access) to have ping support. If this is a database thing, I agree we could only define it once.

gabrielcorado commented on lib/srv/alpnproxy/conn.go:
```diff
@@ -87,3 +89,109 @@ func (conn readOnlyConn) RemoteAddr() net.Addr               { return &utils.Net
+}
+
+func (c *pingConn) Read(p []byte) (int, error) {
+	c.muRead.Lock()
+	defer c.muRead.Unlock()
```
It is for the case we have concurrent `Read` calls, I know it is not that common, but it seems safer to handle those scenarios. Without the lock, we would have data race for `bytesRead` and `currentSize`. This could cause the `Read` calls to try to read ping packets where they would need to read for data content and vice-versa.

There is a `ConcurrentRead` test case to cover those scenarios.

gabrielcorado commented on tool/tsh/proxy.go:
```diff
@@ -448,10 +448,16 @@ func mkLocalProxy(ctx context.Context, opts localProxyOpts) (*alpnproxy.LocalPro
 		return nil, trace.Wrap(err)
 	}
+
+	protocols := append([]alpncommon.Protocol{alpnProtocol}, opts.protocols...)
+	if alpncommon.HasPingSupport(alpnProtocol) {
+		protocols = append([]alpncommon.Protocol{alpncommon.ProtocolWithPing(alpnProtocol)}, protocols...)
+	}
```
Currently, we're letting callers define the protocols list without modifying it or validating it, and I just wanted to keep it the same behavior. In addition, I'm considering the ping protocol an "opt-in" feature, where you should specify its usage. That said, it makes more sense to let callers know about it instead of hiding it inside the `alpnproxy` package.

In future releases, if we decide that this is the default behavior, I agree that it should be transparent for callers. (But in this case, we might end up removing the suffix and making the databases protocols already act as a ping protocol).

greedy52 commented on lib/srv/alpnproxy/conn_test.go:
```diff
@@ -0,0 +1,361 @@
+			// Place into a goroutine to avoid freezing tests in case where the
+			// write succeeds.
+			go func() {
+				largeSlice := make([]byte, math.MaxUint32+1)
```
is this 4g of memory? i would not dare to run this test.

gabrielcorado replied:
Yeah, these tests (with large inputs) were not intended to be executed with the unit testing suite. I considered:
1. removing them;
2. keeping them but only running when a flag/env var is provided;

What do you think?

greedy52 replied:
how about make it configurable so test ping conn can use a different default?

gabrielcorado replied:
"Regular sized" messages are well covered on other tests. These tests were done specifically to test these edge cases where the connections are exchanging this large-sized message. I'll add a `t.Skip` for now to give the PR unit tests a chance to complete, and we can decide if we want to keep them.

jakule replied:
I don't think we should allow that scenario in any case, especially in production. If you want to exchange more than 20MB probably, you should divide the message into chunks. If you run Teleport in a memory-restricted environment, then allocating a few GB of memory will probably kill the process because of OOM.
I suggest changing the max size of a single package to, let's say, 1MB, reduce the size here, and run this test always.

greedy52 commented on lib/srv/alpnproxy/conn_test.go:
```diff
@@ -0,0 +1,361 @@
+
+	// Dial routine.
+	go func() {
+		conn, err := net.Dial("tcp", listener.Addr().String())
```
is it possible to use something like `net.Pipe` instead of dial?

gabrielcorado replied:
Most of the tests are using `net.Pipe` connections. Only one test uses TCP connections to test the "asynchronous" Read/Write, where everything is written to the connection and later is consumed. I've tried to do this with the `bufio` package, but it leads to implementing a `net.Conn` with buffering inside. It felt a lot more code was written just to test this specific scenario, where TCP connections provide everything needed.

greedy52 replied:
cool. if it's a lot of effort then not worth it

just curious if this would work together with a buffer?
https://github.com/gravitational/teleport/blob/edc29051ec17aead90a6a76cabd43446472d2afc/lib/utils/pipenetconn.go#L26-L41

gabrielcorado replied:
I've tried this connection with `bufio` Reader/Writer, and it didn't work well. But thinking now, using just regular `bytes.Buffer` with it might do the job.

gabrielcorado replied:
It worked well. Thanks again!

greedy52 commented on lib/srv/alpnproxy/local_proxy.go:
```diff
@@ -158,7 +158,16 @@ func (l *LocalProxy) handleDownstreamConnection(ctx context.Context, downstreamC
 	}
-	defer upstreamConn.Close()
+	defer tlsConn.Close()
+	if err := tlsConn.Handshake(); err != nil {
```
is `Handshake` required?

gabrielcorado replied:
This call guarantees that we have the `NegotiatedProtocol`  from the connection state (used to define if the connection needs to be wrapped or not) before performing any Read/Write.

greedy52 replied:
ok. i could be wrong but i thought `Dial` already handshakes

gabrielcorado replied:
I can do a quick test to guarantee it is necessary, but as the docs state:

> Handshake runs the client or server handshake protocol if it has not yet been run.
Most uses of this package need not call Handshake explicitly: the first Read or Write will call it automatically.

gabrielcorado replied:
Yeah, I confirmed (with a test and looking at the code) that Dial performs the `Handshake`. I'll remove it. Thanks for pointing it out.

greedy52 commented on tool/tsh/proxy.go:
```diff
@@ -450,10 +450,16 @@ func mkLocalProxy(ctx context.Context, opts localProxyOpts) (*alpnproxy.LocalPro
+
+	protocols := append([]alpncommon.Protocol{alpnProtocol}, opts.protocols...)
+	if alpncommon.HasPingSupport(alpnProtocol) {
+		protocols = append([]alpncommon.Protocol{alpncommon.ProtocolWithPing(alpnProtocol)}, protocols...)
```
nitpick
```suggestion
		protocols = append(alpncommon.ProtocolsWithPing(alpnProtocol), protocols...)
```

gabrielcorado commented on lib/service/service.go:
```diff
@@ -3654,18 +3654,18 @@ func (process *TeleportProcess) initProxyEndpoint(conn *Connector) error {
 				HandlerWithConnInfo: alpnproxy.ExtractMySQLEngineVersion(dbProxyServer.MySQLProxy().HandleConnection),
```
> Maybe worth having a test for this case where there is an extra mysql engine version in the proto list.

Do you mean having both (MySQL + Ping suffix and MySQL version)? If yes, not sure if it is necessary since the PR doesn't change any of the router/mux.

> on second thought, have we considered similar approach for ping protocol (like using an extra protocol) instead of doubling each protocol? curious what pros and cons comparing current approach

[Yeah, It was discussed on the RFD if you want to take a look.](https://github.com/gravitational/teleport/pull/15152#discussion_r938097049)

greedy52 replied:
cool. That makes sense. So it will break if new client -> old server, cuz client can't tell if server multiplex ping or not

gabrielcorado commented on lib/srv/alpnproxy/common/protocols.go:
```diff
@@ -132,10 +141,64 @@ func ToALPNProtocol(dbProtocol string) (Protocol, error) {
+}
+
+// protocolListContains checks if a protocol is present on the list.
+func protocolListContains(protocols []Protocol, protocol Protocol) bool {
```
@smallinsky, are we going to backport it to v9?

smallinsky replied:
Yea, We should backport this to v9

gabrielcorado replied:
Do you mean having this PR with the `Contains` function and the backport with this version?

gabrielcorado committed:
Apply suggestions from code review

Co-authored-by: STeve (Xin) Huang <xin.huang@goteleport.com>

gabrielcorado committed:
refactor(alpnproxy): use single counter to track ping conn reads

gabrielcorado committed:
refactor(alpnproxy): remove unnecessary `Handshake` call

gabrielcorado commented on lib/srv/alpnproxy/conn.go:
```diff
@@ -87,3 +91,114 @@ func (conn readOnlyConn) RemoteAddr() net.Addr               { return &utils.Net
+		return 0, err
+	}
+
+	// Check if the current size is larger than the provided buffer.
+	readSize := c.currentSize
+	if c.currentSize > uint32(len(p)) {
+		readSize = uint32(len(p))
+	}
```
Changed to use a single counter. Thanks for the suggestion.

gabrielcorado committed:
tests(alpnproxy): use PipedConnection instead of TCP connections

gabrielcorado committed:
tests(alpnproxy): disable large content tests by default

gabrielcorado committed:
refactor(types): add missing omitempty

gabrielcorado committed:
chore(alpnproxy): add missing godoc

jakule commented on lib/srv/alpnproxy/conn_test.go:
```diff
@@ -0,0 +1,334 @@
+			defer cancel()
+
+			r, w := makePingConn(t)
+			defer r.Close()
+			defer w.Close()
```
```suggestion
			t.Cleanup(r.Close)
			t.Cleanup(w.Close)
```

jakule commented on lib/srv/alpnproxy/proxy.go:
```diff
@@ -379,14 +399,49 @@ func (p *Proxy) handleConn(ctx context.Context, clientConn net.Conn) error {
+		defer ticker.Stop()
+		for {
+			select {
+			case <-ctx.Done():
```
If we're "done" here, shouldn't we also close the connection?

gabrielcorado replied:
I was following the ["recommendation" from the `Serve` function to let the handler close the connection](https://github.com/gravitational/teleport/blob/master/lib/srv/alpnproxy/proxy.go#L310). Not sure if we can have some issue with database connections tho, but I'm assuming the handlers are closing it properly.

greedy52 replied:
`ctx` here is the server context. Seems with this implementation, we are mostly relying on `pingConn.WritePing()` to fail when the connection is closed. Not sure if it is more convenient to built-in this periodic ping function inside `pingConn` so the `pingConn` can track Close better and stop the ticker.

gabrielcorado replied:
It might be redundant to have `ctx.Done` in this situation. Since the connection will also be closed, I can drop it too.

> Not sure if it is more convenient to built-in this periodic ping function inside `pingConn` so the `pingConn` can track Close better and stop the ticker.

I think it is better to let this outside the "protocol implementation", allowing the callers/users to decide how they will use the `WritePing` function. For example, the Local Proxy doesn't send any ping information and uses the connection.

jakule commented on lib/srv/alpnproxy/proxy_test.go:
```diff
@@ -505,3 +530,69 @@ func TestMatchMySQLConn(t *testing.T) {
 }
+
+func TestProxyPingConnections(t *testing.T) {
+	log.SetLevel(log.DebugLevel)
+	log.SetFormatter(utils.NewDefaultTextFormatter(trace.IsTerminal(os.Stderr)))
```
Why this test makes some modifications to the global logger?

gabrielcorado replied:
Nice catch. It was left from my debug.

jakule commented on lib/config/fileconf_test.go:
```diff
@@ -137,6 +139,19 @@ func TestAuthSection(t *testing.T) {
+				cfg["auth_service"].(cfgMap)["proxy_ping_interval"] = "10s"
+			},
+			expectError:             require.NoError,
+			expectProxyPingInterval: requireEqual(types.Duration(10 * time.Minute)),
```
Maybe I'm missing something, but `proxy_ping_interval` is set to 10 SECONDS, and here we expect 10 MINUTES?

gabrielcorado replied:
You're right. The `expect` part was missing. Thanks!

jakule commented on lib/service/service_test.go:
```diff
@@ -495,6 +495,12 @@ func TestSetupProxyTLSConfig(t *testing.T) {
 				"h2",
 				"http/1.1",
 				"acme-tls/1",
+				"teleport-postgres-ping",
```
Do we really need to duplicate every single entry here? Sounds like something that is very easy to miss when adding a new database protocol and very hard to detect. Could we, for example, generate all combinations by appending `-ping` on init for example?

jakule commented on lib/srv/alpnproxy/conn_test.go:
```diff
@@ -0,0 +1,334 @@
+			defer r.Close()
+			defer w.Close()
+
+			largeSlice := make([]byte, math.MaxUint32)
```
We should not allocate 4GB in tests in any scenario. I don't think we should go over our `MaxHTTPRequestSize` for any reason: https://github.com/gravitational/teleport/blob/808be5be3b664d43cbf50f7ed55c3f90263feef4/constants.go#L500

gabrielcorado committed:
Apply suggestions from code review

Co-authored-by: Jakub Nyckowski <jakub.nyckowski@goteleport.com>

gabrielcorado committed:
test(alpnproxy): remove unnecessary logger

gabrielcorado commented on lib/srv/alpnproxy/conn.go:
```diff
@@ -87,3 +91,106 @@ func (conn readOnlyConn) RemoteAddr() net.Addr               { return &utils.Net
+	defer c.muWrite.Unlock()
+
+	// Avoid casting into something bigger than acceptable.
+	if len(p) > math.MaxUint32 {
```
Brief context: I've tried to break the connection using database dumps up to 5GB, but without success (their protocol handled it differently, usually with streaming). I've added this as a double-checker if any underlying protocol doesn't take it properly. It is also worth noting that this is a concern only with `Database -> Teleport` communications as it breaks the entire agent if not handled correctly.

With that being said, I think having a more reasonable limit will also handle this. What do you think, @greedy52 @smallinsky?

greedy52 replied:
How about having a `MaxWriteChunk` say default to ~20MB (MySQL default max packet is 16MB, Postgres is 20MB), and then if `len(p)` is greater than the chunk, instead of return an error, break `p` into chunks and then send each chunk in a loop?

smallinsky replied:
I would rather try to avoid this kind of logic in `pingConn` and try to keep things simple and relay on TCP MTU instead.

smallinsky replied:
For some database write protocols the max size of the package is quite large  for instance mysql says that max limit for single package is set to [1GB](https://dev.mysql.com/doc/refman/8.0/en/packet-too-large.html#:~:text=The%20largest%20possible%20packet%20that,server%20or%20client%20is%201GB.) 
> The largest possible packet that can be transmitted to or from a MySQL 8.0 server or client is 1GB.

smallinsky replied:
I see but this limitation should be enforced on db engine db proxy side where database packets are handled (this pingConn handler doesn't do any allocation). It is hard to anticipate what max packet limit size will be for database protocols there is always a risk that we will add a protocol that by default has a large limit like 64MB

gabrielcorado replied:
This check would only guarantee we're not panicking due to an invalid cast (which nothing would directly prevent from happening). I didn't consider it a problem since we're not copying data or doing any allocation (or at least it is not done on the `PingConn`), so limiting it here would not solve any potential issue.

If you feel like this is an unnecessary check, we can drop it. Still, I wouldn't give it the purpose of limiting what size can be sent. This limitation is particular with each database protocol we implement, so it seems better to deal with it per database protocol.

For curiosity, in my tests, I found that having those large packets would fail a step before, while generating the audit events:
```
2022-08-19T13:13:21-03:00 ERRO [DB:SERVIC] Audit write timed out after 5s. Will be losing events for the next 30s. events/auditwriter.go:361
2022-08-19T13:13:57-03:00 ERRO             Failed to emit audit event. error:[
ERROR REPORT:
Original Error: trace.aggregate grpc: received message larger than max (16777441 vs. 4194304)
Stack Trace:
        /teleport/lib/events/emitter.go:289 github.com/gravitational/teleport/lib/events.(*MultiEmitter).EmitAuditEvent
        /teleport/lib/events/emitter.go:168 github.com/gravitational/teleport/lib/events.(*CheckingEmitter).EmitAuditEvent
        /teleport/lib/events/emitter.go:94 github.com/gravitational/teleport/lib/events.(*AsyncEmitter).forward
        /opt/homebrew/Cellar/go/1.18.3/libexec/src/runtime/asm_arm64.s:1263 runtime.goexit
User Message: grpc: received message larger than max (16777441 vs. 4194304)] events/emitter.go:170
```

jakule replied:
@gabrielcorado Thanks for testing that. In this case, I'd leave it as it is now; just add a comment that we limit the package size at the GRPC level.

gabrielcorado committed:
test(config): add missing expect on config

smallinsky commented on lib/srv/alpnproxy/proxy.go:
```diff
@@ -95,6 +97,21 @@ func MatchByProtocol(protocols ...common.Protocol) MatchFunc {
+// MatchByProtocolWithPing creates match function based on client TLS APLN
+// protocol matching also their ping protocol variations.
+func MatchByProtocolWithPing(protocols ...common.Protocol) MatchFunc {
+	m := make(map[common.Protocol]struct{})
+	for _, v := range protocols {
+		m[common.ProtocolWithPing(v)] = struct{}{}
+		m[v] = struct{}{}
+	}
+
+	return func(_, alpn string) bool {
+		_, ok := m[common.Protocol(alpn)]
+		return ok
+	}
```
Can we reduce this to something like: 
```suggestion
	return MatchByProtocol(append(protocol, ProtocolsWithPing(protocols)...))
```

gabrielcorado committed:
Update lib/srv/alpnproxy/proxy.go

Co-authored-by: Marek Smoliński <marek@goteleport.com>

gabrielcorado committed:
refactor(alpnproxy): use DatabaseProtocols in the supported protocols

gabrielcorado committed:
fix: change database proxy to use utils.TLSConn instead of *tls.Conn

Previously, when using DB TLS, the database proxies relied on *tls.Conn
to do the Authorize call. With the addition of `*pingConn` the started
to failed, so we needed to add an interface (called `TLSConn`) so that
both connection types will work within this scenario.

gabrielcorado committed:
chore(alpnproxy): fix lint

gabrielcorado committed:
test(service): fix protocols order

greedy52 commented on lib/srv/db/mysql/proxy.go:
```diff
@@ -194,11 +195,14 @@ func (p *Proxy) performHandshake(conn *multiplexer.Conn, server *server.Conn) (*
 	case *multiplexer.Conn:
-		tlsConn, ok := c.Conn.(*tls.Conn)
-		if !ok {
+		switch tlsConn := c.Conn.(type) {
+		case *tls.Conn:
+			return tlsConn, nil
+		case *alpnproxy.PingConn:
+			return tlsConn, nil
```
should we type check against `utils.TLSConn` vs the actual types?

gabrielcorado committed:
refactor(mysql): match connection with TLSConn

gabrielcorado committed:
test(alpnproxy): remove skipped tests

gabrielcorado committed:
Merge branch 'master' into gabrielcorado/tls-multiplex

gabrielcorado committed:
Merge branch 'master' into gabrielcorado/tls-multiplex

smallinsky commented on lib/srv/alpnproxy/conn.go:
```diff
@@ -87,3 +92,109 @@ func (conn readOnlyConn) RemoteAddr() net.Addr               { return &utils.Net
+	return nil
+}
+
+// Write writes provided content to the underlaying connection with proper
```
```suggestion
// Write writes provided content to the underlying connection with proper
```

smallinsky commented on lib/srv/db/sqlserver/proxy.go:
```diff
@@ -41,7 +41,7 @@ type Proxy struct {
 // HandleConnection accepts connection from a SQL Server client, authenticates
 // it and proxies it to an appropriate database service.
-func (p *Proxy) HandleConnection(ctx context.Context, proxyCtx *common.ProxyContext, tlsConn *tls.Conn) error {
+func (p *Proxy) HandleConnection(ctx context.Context, proxyCtx *common.ProxyContext, tlsConn utils.TLSConn) error {
```
I don't see where the `tls.Conn`  is required for below logic we can just use net.Conn here.

gabrielcorado committed:
refactor(sqlserver): change TLSConn to net.Conn

gabrielcorado committed:
Merge branch 'master' into gabrielcorado/tls-multiplex

gabrielcorado committed:
Merge branch 'master' into gabrielcorado/tls-multiplex

gabrielcorado committed:
fix(alpnproxy): match raw protocols in ping

gabrielcorado committed:
Merge branch 'master' into gabrielcorado/tls-multiplex

gabrielcorado committed:
Merge branch 'gabrielcorado/tls-multiplex' of github.com:gravitational/teleport into gabrielcorado/tls-multiplex

gabrielcorado committed:
Merge branch 'master' into gabrielcorado/tls-multiplex
